spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -1233,6 +1233,9 @@ public void run() {
 				catch (Exception e) {
 					handleConsumerException(e);
 				}
+				finally {
+					finishInvoke();
```

Comment:
I wanted it to run after `handleConsumerException()` so the trace context will be available for an exception outside of the listener processing e.g. when offsets fail to commit.

Reply:
Right, but as I said in my comment on the issue; it needs to be called conditionally in the finally block - only if an exception was thrown; otherwise it will be called twice.

Reply:
You said:

> We need to handle the `poll()` throwing an exception, with a conditional call to `finishInvoke()` in a `finally` block in the `run()` method. e.g. a boolean `success` set to false just after the `while()` and true after the `pollAndInvoke()` and use it in the `finally` block to call the clean up if an exception occurred.

My understanding of the tracing use-case that started this PR is that `finishInvoke()` must be run if `beforePoll()` was run. Whether or not records were returned from `poll()`, or processed successfully by listener, or exception was thrown somewhere, is not important. The only goal is to run `finishInvoke()` after `beforePoll()`.

Can you please explain what do you mean by:

>  clean up if an exception occurred.

because it feels like the `finishInvoke()` cleanup is not dependent on an exception being thrown during processing. Unless you are suggesting that `finishInvoke()` should only run if `beforePoll()` run without throwing an exception? My apologies, I'm clearly missing something.

Reply:
Well, I was thinking that we should call `finishInvoke()` after each record in a record listener but I guess you are right; these methods should be just scoped around the `poll()` and completion of processing all records from the poll.

In that case, we should rename `finishInvoke` to something like `recordsProcessed()` to decouple it from invoking the listener.

Most likely record interceptors won't use them, but we should provide them for consistency.

Reply:
I renamed `finishInvoke()` to `afterRecordsProcessed()` and improved the Javadoc. 

I still have to figure out how to test it because `beforePoll()` and `afterRecordsProcessed()` is called in a loop outside of the test. I plan to look into it over the weekend.